### PR TITLE
125 rework controller changes

### PR DESCRIPTION
**Pull request by gdoffe**
Improve controllers chain management with fixes and new features.

### CODE REVIEW

**Review by ecourtois — CHANGES_REQUESTED** [review]
Commit titles are too long for bff42053, 3f08e85 and 62d6236.
Typo in commit title 9712717.
Commit 62d6236 must be inserted before 9712717.

See also questions/comments in code.

File: motion_control/motion_control_common/include/motion_control_common/MetaController.hpp [hunk 1]
```diff
@@ -90,6 +90,48 @@ class MetaController :
...
+        BaseController *ctrl    ///< [in]  new controller to add
+        ) override {
+        if (! ctrl->set_meta(this)) {
+            return;
+        }
+        if (INPUT_SIZE != ctrl->nb_inputs()) {
+            COGIP_DEBUG_CERR("Error: First controller must have the same number of inputs (" << ctrl->nb_inputs() << ") as the meta controller (" << INPUT_SIZE << ").");
+            return;
+        }
+        this->insert(this->end(), ctrl);
```
[the remark this answers is not in the document]
Comment (gdoffe) [reply motion_control/motion_control_common/include/motion_control_common/MetaController.hpp:105]: `etl::deque` is exactly what I need, thx.

File: motion_control/motion_control_common/include/motion_control_common/ParallelMetaController.hpp [hunk 2]
```diff
@@ -68,6 +68,47 @@ class ParallelMetaController :
...
+
+        BaseController *ctrl_to_replace = *ctrl_iterator;
+
+        if ((ctrl_to_replace->nb_inputs() != ctrl->nb_inputs()) ||
+            (ctrl_to_replace->nb_outputs() != ctrl->nb_outputs())) {
+            return;
+        }
+
+        ctrl_iterator = this->erase(ctrl_iterator);
+        this->insert(--ctrl_iterator, ctrl);
```
[the remark this answers is not in the document]
Comment (gdoffe) [reply motion_control/motion_control_common/include/motion_control_common/ParallelMetaController.hpp:109]: no reason, just an oversight ! done.

File: motion_control/motion_control_common/include/motion_control_common/ParallelMetaController.hpp [hunk 3]
```diff
@@ -68,6 +68,47 @@ class ParallelMetaController :
...
     };
 
+    /// Add a controller to the beginning of the controllers chain.
+    void insert_controller(
+        BaseController *ctrl    ///< [in]  new controller to add
+        ) override {
+        if (! ctrl->set_meta(this)) {
+            return;
+        }
+        this->insert(this->end(), ctrl);
```
[the remark this answers is not in the document]
Comment (gdoffe) [reply motion_control/motion_control_common/include/motion_control_common/ParallelMetaController.hpp:78]: done

File: motion_control/motion_control_common/include/motion_control_common/MetaController.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (gdoffe) [reply motion_control/motion_control_common/include/motion_control_common/MetaController.hpp:105]: done

**Review by ecourtois — APPROVED** [review]

**Pull request merged**